scratchpad/authentiacation.ts

```diff
@@ -0,0 +1,32 @@
+
+export class Authenticate {
+  constructor(
+    private @inject('authentication.required') required,
+    private @inject('authentication.user') user
+  )
+  static name = 'authentication.user';
+  value() {
+    console.log(this.user); // {username: ...};
+    return () => {
+        const isAuthRequired = this.required;
+        const user = this.user;
+
+        if (isAuthRequired && !user) {
+          throw ...;
+        }
+    }
+  }
+}
```

Comment:
While this factory-class based approach is ok, I think it may be a bit cumbersome when writing extensions in JavaScript. I am proposing to add support for a second way of defining bindings:

```js
export const binding = {
  name: 'authentication.user',
  inject: [
    'authentication.required',
    'authentication.user',
  ],
  value: (isAuthRequired, user) => {
    console.log(user);
    return () => {
      if (isAuthRequired && !user) {
        throw ...;
      }
    };
  },
}
```

Because the items in `inject` array cannot use aliased decorators directly (e.g. `injectAuthenticatedUser` instead of `authentication.user`), I am proposing that all aliased decorators should have a `key` property containing the alias name (binding key).

```js
export const binding = {
  name: 'authentication.user',
  inject: [
    injectAuthenticationRequired.key,
    injectAuthenticatedUser.key,
  ],
  value: (isAuthRequired, user) => {
    console.log(user);
    return () => {
      if (isAuthRequired && !user) {
        throw ...;
      }
    };
  },
}

// elsewhere
export function injectAuthenticatedUser() {
  return inject('authentication.user');
}
injectAuthenticatedUser.key = 'authentication.user';
```

I think this is an additive change than can be left out of scope of this initial work. @ritch let me know your thoughts, should I create a new follow-up issue for this?